dll/win32/framedyn/WbemTime.cpp

```diff
@@ -0,0 +1,522 @@
+WBEMTimeSpan::WBEMTimeSpan(
+    int iDays,
+    int iHours,
+    int iMinutes,
+    int iSeconds,
+    int iMSec,
+    int iUSec,
+    int iNSec
+)
+{
+    m_Time = 86400 * iDays + 3600 * iHours + 60 * iMinutes + iSeconds;
+    m_Time *= 10000000;
+    m_Time += 10000 * iMSec + 10 * iUSec + iNSec / 100;
+}
+
+// add operator adds one time span to another, placing the sum in a new WBEMTimeSpan object returned by the method.
+WBEMTimeSpan WBEMTimeSpan::operator+ (const WBEMTimeSpan &uAdd) const
+{
+    WBEMTimeSpan rt;
+
+    if (IsOk() && uAdd.IsOk())
+    {
+        rt.m_Time = m_Time + uAdd.m_Time;
+    }
+
+    return rt;
+}
+
+// add-and-assign operator adds one time span to another. The operation create a new time span that contains the resulting time.
+const WBEMTimeSpan& WBEMTimeSpan::operator+= (const WBEMTimeSpan &uAdd)
+{
+    if (IsOk() && uAdd.IsOk())
+    {
+        m_Time += uAdd.m_Time;
+    }
+    else
+    {
+        m_Time = INVALID_TIME;
+    }
+
+    return *this;
+}
+
+// subtract operator (�) subtracts a time span from the object on which the method is executed. The result is a new time span that contains the difference in time.
+WBEMTimeSpan WBEMTimeSpan::operator- (const WBEMTimeSpan &uSub) const
+{
+    WBEMTimeSpan rt;
+
+    if (IsOk() && uSub.IsOk() && m_Time >= uSub.m_Time)
+    {
+        rt.m_Time = m_Time - uSub.m_Time;
+    }
+
+    return rt;
+}
+
+// subtract and assign operator (�=).
+const WBEMTimeSpan& WBEMTimeSpan::operator-= (const WBEMTimeSpan &uSub)
+{
+    if (IsOk() && uSub.IsOk() && m_Time >= uSub.m_Time)
+    {
+        m_Time -= uSub.m_Time;
+    }
+    else
+    {
+        m_Time = INVALID_TIME;
+    }
+
+    return *this;
+}
+
+// Converts a BSTR time interval value to a WBEMTimeSpan object in CIM date and time format.
+const WBEMTimeSpan& WBEMTimeSpan::operator= (const BSTR bstrDMTFFormat)
+{
+    m_Time = INVALID_TIME;
+
+    if (bstrDMTFFormat != NULL && wcslen(bstrDMTFFormat) == 25 && bstrDMTFFormat[14] == L'.' && bstrDMTFFormat[21] == L':')
+    {
+        int iDays, iHours, iMinutes, iSeconds, iUSec;
+        if (swscanf(bstrDMTFFormat, L"%8d%2d%2d%2d.%6d:000", &iDays, &iHours, &iMinutes, &iSeconds, &iUSec) == 5)
+        {
+            WBEMTimeSpan t(iDays, iHours, iMinutes, iSeconds, 0, iUSec, 0);
+            m_Time = t.m_Time;
+        }
+    }
+
+    return *this;
+}
+
+// gets the time span as a BSTR in Date and Time format.
+BSTR WBEMTimeSpan::GetBSTR(void) const throw (CHeap_Exception)
+{
+    if (!IsOk())
+    {
+        return NULL;
+    }
+
+    ULONGLONG totalSeconds = m_Time / 10000000;
+    ULONGLONG iDays = totalSeconds / 86400;
+    int iHours = (totalSeconds % 86400) / 3600;
+    int iMinutes = (totalSeconds % 3600) / 60;
+    int iSeconds = totalSeconds % 60;
+    int iUSec = (m_Time % 10000000) / 10;
+
+    BSTR v = (BSTR)SysAllocStringLen(NULL, 26);
+    swprintf(v, L"%08I64i%02d%02d%02d.%06ld:000", iDays, iHours, iMinutes, iSeconds, iUSec);
```

Comment:
There aren't any printf member function for BSTR type? Also, is it  possible to use secure-string functions here? (StringCchPrintfW for example).